File: error-notebook/scripts/validate-public-repo.mjs
#!/usr/bin/env node
/** 校验公开仓库不包含个人错题、敏感凭据或真实本机路径。 */

import fs from "node:fs";
import path from "node:path";
import { spawnSync } from "node:child_process";

import { parseNotebook } from "./notebook-parser.mjs";
import { bundledNotebook, skillRoot } from "./notebook-paths.mjs";


const repositoryRoot = path.dirname(skillRoot);
const forbiddenPatterns = new Map([
  ["Windows 用户目录", /[A-Za-z]:\\Users\\(?!<)[^\\\s`]+/u],
  ["本机代码目录", /[A-Za-z]:\\(?:code|workspace)\\[^\s`]+/iu],
  [
    "内网 IPv4 地址",
    /\b(?:10\.\d{1,3}\.\d{1,3}\.\d{1,3}|192\.168\.\d{1,3}\.\d{1,3}|172\.(?:1[6-9]|2\d|3[01])\.\d{1,3}\.\d{1,3})\b/u,
  ],
  ["私钥", /-----BEGIN (?:RSA |EC |OPENSSH )?PRIVATE KEY-----/u],
  ["GitHub Token", /\b(?:gh[opusr]_[A-Za-z0-9]{20,}|github_pat_[A-Za-z0-9_]{20,})\b/u],
  ["OpenAI API Key", /\bsk-(?:proj-)?[A-Za-z0-9_-]{20,}\b/u],
  ["AWS Access Key", /\bAKIA[0-9A-Z]{16}\b/u],
]);

/** 返回已跟踪文件和未忽略的新文件，供提交前检查。 */
function repositoryFiles() {
  const result = spawnSync(
    "git",
    ["ls-files", "-z", "--cached", "--others", "--exclude-standard"],
    { cwd: repositoryRoot, encoding: "buffer" },
  );
  if (result.error) {
    throw result.error;
  }
  if (result.status !== 0) {
    throw new Error(result.stderr.toString("utf8").trim() || "git ls-files 执行失败");
  }
  return result.stdout
    .toString("utf8")
    .split("\0")
    .filter(Boolean)
    .map((item) => path.join(repositoryRoot, item));
}

/** 尝试按严格 UTF-8 解码；二进制或非 UTF-8 文件返回 null。 */
function readUtf8(pathname) {
  try {
    return new TextDecoder("utf-8", { fatal: true }).decode(
      fs.readFileSync(pathname),
    );
  } catch {
    return null;
  }
}

/** 执行公开仓库隐私检查。 */
function main() {
  const errors = [];
  const templateText = fs.readFileSync(bundledNotebook, "utf8");
  const { entries: templateEntries } = parseNotebook(templateText);
  if (templateEntries.length > 0) {
    errors.push(
      `公开模板包含 ${templateEntries.length} 条真实经验；请迁移到私有数据文件。`,
    );
  }

  for (const pathname of repositoryFiles()) {
    const text = readUtf8(pathname);
    if (text === null) {
      continue;
    }
    const relativePath = path.relative(repositoryRoot, pathname);
    for (const [patternName, pattern] of forbiddenPatterns) {
      if (pattern.test(text)) {
        errors.push(`${relativePath} 检测到${patternName}。`);
      }
    }
  }

  if (errors.length > 0) {
    console.error("公开仓库校验失败：");
    for (const error of errors) {
      console.error(`- ${error}`);
    }
    return 1;
  }
  console.log(
    "公开仓库校验通过：公开模板为空，已跟踪文件未发现高风险凭据或个人绝对路径。",
  );
  return 0;
}


process.exitCode = main();

File: error-notebook/scripts/validate_public_repo.py
#!/usr/bin/env python3
"""校验公开仓库不包含个人错题、敏感凭据或真实本机路径。"""

from __future__ import annotations

from pathlib import Path
import re
import subprocess
import sys

from notebook_paths import BUNDLED_NOTEBOOK, SKILL_ROOT
from search_notebook import parse_notebook


REPOSITORY_ROOT = SKILL_ROOT.parent
FORBIDDEN_PATTERNS = {
    "Windows 用户目录": re.compile(r"[A-Za-z]:\\Users\\(?!<)[^\\\s`]+"),
    "本机代码目录": re.compile(r"[A-Za-z]:\\(?:code|workspace)\\[^\s`]+", re.IGNORECASE),
    "内网 IPv4 地址": re.compile(
        r"\b(?:10\.\d{1,3}\.\d{1,3}\.\d{1,3}|"
        r"192\.168\.\d{1,3}\.\d{1,3}|"
        r"172\.(?:1[6-9]|2\d|3[01])\.\d{1,3}\.\d{1,3})\b"
    ),
    "私钥": re.compile(r"-----BEGIN (?:RSA |EC |OPENSSH )?PRIVATE KEY-----"),
    "GitHub Token": re.compile(
        r"\b(?:gh[opusr]_[A-Za-z0-9]{20,}|github_pat_[A-Za-z0-9_]{20,})\b"
    ),
    "OpenAI API Key": re.compile(r"\bsk-(?:proj-)?[A-Za-z0-9_-]{20,}\b"),
    "AWS Access Key": re.compile(r"\bAKIA[0-9A-Z]{16}\b"),
}


def repository_files() -> list[Path]:
    """返回已跟踪文件和未忽略的新文件，供提交前检查。"""

    result = subprocess.run(
        ["git", "ls-files", "-z", "--cached", "--others", "--exclude-standard"],
        cwd=REPOSITORY_ROOT,
        check=True,
        capture_output=True,
    )
    return [
        REPOSITORY_ROOT / item.decode("utf-8")
        for item in result.stdout.split(b"\0")
        if item
    ]


def main() -> int:
    """执行公开仓库隐私检查。"""

    errors: list[str] = []
    template_text = BUNDLED_NOTEBOOK.read_text(encoding="utf-8")
    _, _, template_entries = parse_notebook(template_text)
    if template_entries:
        errors.append(
            f"公开模板包含 {len(template_entries)} 条真实经验；请迁移到私有数据文件。"
        )

    for path in repository_files():
        try:
            text = path.read_text(encoding="utf-8")
        except (UnicodeDecodeError, OSError):
            continue
        relative_path = path.relative_to(REPOSITORY_ROOT)
        for pattern_name, pattern in FORBIDDEN_PATTERNS.items():
            if pattern.search(text):
                errors.append(f"{relative_path} 检测到{pattern_name}。")

    if errors:
        print("公开仓库校验失败：", file=sys.stderr)
        for error in errors:
            print(f"- {error}", file=sys.stderr)
        return 1

    print("公开仓库校验通过：公开模板为空，已跟踪文件未发现高风险凭据或个人绝对路径。")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
